File: strategy-builder/components/agent/csharp/microsoft-agent-framework/AgentSetup.cs
/// <summary>
/// Agent setup — how to configure the Microsoft Agent Framework (MAF)
/// on Azure AI Foundry using MAF Workflows.
///
/// The pattern:
///   1. Create a ResponsesClient pointing at Azure OpenAI (or mock)
///   2. Create three discrete specialist agents
///   3. Attach local knowledge tools and resolution tools to each specialist
///   4. Build one workflow executor per specialist
///   5. Create a CheckpointManager for multi-turn conversation state
///
/// Architecture diagram (MAF Workflow with discrete specialists):
///
///     User message
///         |
///         v
///   ┌───────────────────────────────────────────────┐
///   │  MAF Workflow (InProcessExecution)             │
///   │                                               │
///   │  Specialist Executor selected by mode         │
///   │  ├─ lookup_*_knowledge  (local tool)          │
///   │  └─ resolve_*           (lambda tool)         │
///   └───────────────────────────────────────────────┘
///         │
///         v
///   WatchStreamAsync() → IAsyncEnumerable&lt;WorkflowEvent&gt;
///   ├─ AgentResponseUpdateEvent (text deltas → message.delta SSE)
///   ├─ AgentResponseUpdateEvent (FunctionCallContent → resolution detection)
///   └─ SuperStepCompletedEvent (checkpoint for multi-turn state)
///
/// The selected specialist is the conversational agent for the current mode.
///
/// Multi-turn chaining: The MAF CheckpointManager captures the full
/// executor state (including the agent's conversation history) after
/// each super-step. On subsequent parleys, ResumeStreamingAsync
/// restores the checkpoint and the conversation continues seamlessly.
/// </summary>

#pragma warning disable OPENAI001 // ResponsesClient is experimental

using System.ComponentModel;
using Microsoft.Agents.AI;
using Microsoft.Agents.AI.Workflows;
using Microsoft.Extensions.AI;
using OpenAI.Responses;

namespace CairaAgent;

// ---------------------------------------------------------------------------
// Agent setup result
// ---------------------------------------------------------------------------

/// <summary>
/// The result of agent setup — everything needed to run the workflow.
/// Returns the per-mode workflow map and checkpoint manager. The workflow engine
/// handles execution, streaming events, and state management.
/// </summary>
public sealed class AgentSetupResult
{
    /// <summary>
    /// Legacy default workflow reference kept for compatibility with callers
    /// that still expect a single workflow instance.
    /// </summary>
    public required Workflow? Workflow { get; init; }

    /// <summary>
    /// In-memory checkpoint manager — stores workflow state (including
    /// conversation history) between parleys. Each conversation gets its
    /// own checkpoint chain via SuperStepCompletedEvent.CompletionInfo.Checkpoint.
    /// </summary>
    public required CheckpointManager CheckpointManager { get; init; }

    /// <summary>
    /// Workflow per activity mode.
    /// </summary>
    public required IReadOnlyDictionary<string, Workflow> WorkflowsByMode { get; init; }
}

// ---------------------------------------------------------------------------
// Agent setup
// ---------------------------------------------------------------------------

public static class AgentSetup
{
    /// <summary>
    /// Resolution tool name prefix — used to detect resolution tool calls
    /// in AgentResponseUpdateEvent.Update.Contents (FunctionCallContent).
    /// </summary>
    internal const string ResolutionToolPrefix = "resolve_";

    /// <summary>
    /// Create the specialist workflows.
    ///
    /// The pattern:
    ///   1. Create specialist agents with focused instructions
    ///   2. Add local knowledge and resolution tools to each specialist
    ///   3. Build one workflow per specialist using BindAsExecutor(emitEvents: true)
    ///   4. Create CheckpointManager.CreateInMemory() for multi-turn state
    /// </summary>
    public static AgentSetupResult Create(AgentConfig config, ILogger logger)
    {
        var responsesClient = CreateResponsesClient(config);

        var sharedInstructions = config.CaptainInstructions.Trim();
        string Compose(string specialistInstructions) => $"{sharedInstructions}\n\n{specialistInstructions}".Trim();

        var shantyAgent = responsesClient.AsAIAgent(
            instructions: Compose(config.ShantyInstructions),
            name: "ShantySpecialist",
            description: "Opportunity discovery specialist for the fictional sales/account-team sample.");

        var treasureAgent = responsesClient.AsAIAgent(
            instructions: Compose(config.TreasureInstructions),
            name: "TreasureSpecialist",
            description: "Account planning specialist for the fictional sales/account-team sample.");

        var crewAgent = responsesClient.AsAIAgent(
            instructions: Compose(config.CrewInstructions),
            name: "CrewSpecialist",
            description: "Account-team staffing specialist for the fictional sales/account-team sample.");

        var shantyKnowledge = AIFunctionFactory.Create(
            ([Description("What discovery signal, qualification detail, or customer need is needed")] string query) =>
                System.Text.Json.JsonSerializer.Serialize(new { items = KnowledgeBase.LookupShanty(query) }),
            "lookup_shanty_knowledge",
            "Retrieve fictional sample discovery guidance and qualification cues.");

        var treasureKnowledge = AIFunctionFactory.Create(
            ([Description("What account-planning priority, risk, or next-step detail is needed")] string query) =>
                System.Text.Json.JsonSerializer.Serialize(new { items = KnowledgeBase.LookupTreasure(query) }),
            "lookup_treasure_knowledge",
            "Retrieve fictional sample account-planning guidance, risks, and milestones.");

        var crewKnowledge = AIFunctionFactory.Create(
            ([Description("What staffing role, coverage level, or qualification detail is needed")] string query) =>
                System.Text.Json.JsonSerializer.Serialize(new { items = KnowledgeBase.LookupCrew(query) }),
            "lookup_crew_knowledge",
            "Retrieve fictional sample staffing roles, coverage guidance, and qualifications.");

        var resolveShanty = AIFunctionFactory.Create(
            (
                [Description("Who won the shanty battle")] string winner,
                [Description("Number of qualification signals reviewed")] int rounds,
                [Description("The single most important customer need or buying signal")] string best_verse
            ) =>
            {
                logger.LogInformation("Discovery flow resolved: {Winner} after {Rounds} signals", winner, rounds);
                return $"Discovery flow resolved: {winner} after {rounds} signals.";
            },
            "resolve_shanty",
            "Call this when the discovery activity concludes. Records the fit signal summary using the existing contract.");

        var resolveTreasure = AIFunctionFactory.Create(
            (
                [Description("Whether the plan should advance now")] bool found,
                [Description("Primary account focus area")] string treasure_name,
                [Description("Next milestone, meeting, or workstream")] string location
            ) =>
            {
                var outcome = found ? "Advance" : "Rework";
                logger.LogInformation("Account plan resolved: {Outcome} \"{TreasureName}\" at {Location}", outcome, treasure_name, location);
                return $"Account plan resolved: {outcome} \"{treasure_name}\" at {location}.";
            },
            "resolve_treasure",
            "Call this when the account-planning activity concludes. Records the planning outcome using the existing contract.");

        var resolveCrew = AIFunctionFactory.Create(
            (
                [Description("The recommended coverage level")] string rank,
                [Description("The recommended owner role")] string role,
                [Description("The fictional account team name")] string ship_name
            ) =>
            {
                logger.LogInformation("Staffing flow resolved: {Rank} {Role} on {ShipName}", rank, role, ship_name);
                return $"Staffing flow resolved: {rank} {role} on {ship_name}.";
            },
            "resolve_crew",
            "Call this when the staffing conversation concludes. Records the staffing recommendation using the existing contract.");

        var shantyAgentWithTools = responsesClient.AsAIAgent(
            instructions: Compose(config.ShantyInstructions),
            name: "ShantySpecialist",
            tools:
            [
                shantyKnowledge,
                resolveShanty,
            ]);

        var treasureAgentWithTools = responsesClient.AsAIAgent(
            instructions: Compose(config.TreasureInstructions),
            name: "TreasureSpecialist",
            tools:
            [
                treasureKnowledge,
                resolveTreasure,
            ]);

        var crewAgentWithTools = responsesClient.AsAIAgent(
            instructions: Compose(config.CrewInstructions),
            name: "CrewSpecialist",
            tools:
            [
                crewKnowledge,
                resolveCrew,
            ]);

        var shantyBinding = shantyAgentWithTools.BindAsExecutor(emitEvents: true);
        var treasureBinding = treasureAgentWithTools.BindAsExecutor(emitEvents: true);
        var crewBinding = crewAgentWithTools.BindAsExecutor(emitEvents: true);

        var workflows = new Dictionary<string, Workflow>
        {
            ["shanty"] = new WorkflowBuilder(shantyBinding)
                .WithOutputFrom(shantyBinding)
                .Build(),
            ["treasure"] = new WorkflowBuilder(treasureBinding)
                .WithOutputFrom(treasureBinding)
                .Build(),
            ["crew"] = new WorkflowBuilder(crewBinding)
                .WithOutputFrom(crewBinding)
                .Build(),
        };

        // ---- Create checkpoint manager ----
        // CheckpointManager.CreateInMemory() stores workflow state (including
        // the agent's full conversation history) between parleys. Each
        // SuperStepCompletedEvent yields a CheckpointInfo that can be used
        // to resume the workflow for the next turn.

        var checkpointManager = CheckpointManager.CreateInMemory();

        logger.LogInformation(
            "Mode-specific MAF workflows initialised (model={Model}, name={AgentName})",
            config.Model, config.AgentName);

        return new AgentSetupResult
        {
            Workflow = workflows["shanty"],
            CheckpointManager = checkpointManager,
            WorkflowsByMode = workflows,
        };
    }

    // ---- Helpers ----

    /// <summary>
    /// Create a ResponsesClient pointing at Azure OpenAI or an APIM gateway.
    ///
    /// Always uses AzureOpenAIClient with a TokenCredential:
    ///   - HTTPS endpoint (production): the runtime-appropriate Azure credential for real
    ///     Azure AD tokens via the cognitiveservices scope
    ///   - HTTP endpoint + SKIP_AUTH (local dev / CI): a static dummy
    ///     token credential — the Azure SDK's AzureTokenAuthenticationPolicy
    ///     has no HTTPS enforcement, so this sends "Bearer dummy" which
    ///     any permissive endpoint accepts
    ///
    /// The agent code path is identical regardless of the target endpoint.
    /// </summary>
    private static OpenAI.Responses.ResponsesClient CreateResponsesClient(AgentConfig config)
    {
        var isHttpEndpoint = config.AzureEndpoint.StartsWith("http://");

        Azure.Core.TokenCredential credential = isHttpEndpoint && config.SkipAuth
            ? new StaticTokenCredential("dummy")
            : CreateAzureCredential();

        var azureClient = new Azure.AI.OpenAI.AzureOpenAIClient(
            new Uri(config.AzureEndpoint),
            credential);
        return azureClient.GetResponsesClient(config.Model);
    }

    private static Azure.Core.TokenCredential CreateAzureCredential()
    {
        var managedIdentityEndpoint = Environment.GetEnvironmentVariable("IDENTITY_ENDPOINT")
            ?? Environment.GetEnvironmentVariable("MSI_ENDPOINT");
        if (!string.IsNullOrWhiteSpace(managedIdentityEndpoint))
        {
            var clientId = Environment.GetEnvironmentVariable("AZURE_CLIENT_ID");
            return string.IsNullOrWhiteSpace(clientId)
                ? new Azure.Identity.ManagedIdentityCredential()
                : new Azure.Identity.ManagedIdentityCredential(clientId);
        }

        return new Azure.Identity.DefaultAzureCredential();
    }
}

// ---------------------------------------------------------------------------
// Static token credential — dummy bearer token for HTTP endpoints
// ---------------------------------------------------------------------------

/// <summary>
/// A TokenCredential that always returns a fixed token string.
/// Used for HTTP endpoints (local dev / CI) where the Azure SDK's
/// the runtime-selected Azure credential cannot acquire real tokens. Equivalent to
/// the TypeScript pattern: () => Promise.resolve('dummy').
/// </summary>
internal sealed class StaticTokenCredential : Azure.Core.TokenCredential
{
    private readonly Azure.Core.AccessToken _token;

    public StaticTokenCredential(string token)
    {
        _token = new Azure.Core.AccessToken(token, DateTimeOffset.MaxValue);
    }

    public override Azure.Core.AccessToken GetToken(
        Azure.Core.TokenRequestContext requestContext, CancellationToken cancellationToken)
        => _token;

    public override ValueTask<Azure.Core.AccessToken> GetTokenAsync(
        Azure.Core.TokenRequestContext requestContext, CancellationToken cancellationToken)
        => new(_token);
}

// ---------------------------------------------------------------------------
// SSE formatting helper
// ---------------------------------------------------------------------------

/// <summary>
/// Format SSE events as "event: {name}\ndata: {json}\n\n" strings.
/// Shared by WorkflowRunner for all SSE event types.
/// </summary>
internal static class SseFormatter
{
    public static string Format(string eventName, object data)
    {
        return $"event: {eventName}\ndata: {System.Text.Json.JsonSerializer.Serialize(data)}\n\n";
    }
}

File: reference-architectures/app/api/typescript/openai-agents-sdk/src/agent-setup.ts
/**
 * Agent setup — how to configure the OpenAI Agent SDK on Azure AI Foundry.
 *
 * This file is the "read this first" entry point for developers. It shows:
 *
 *   1. How to connect the SDK to Azure OpenAI via the runtime-appropriate Azure credential
 *   2. How to create agents with system instructions
 *   3. How to attach local knowledge tools and resolution tools to specialists
 *   4. How to pick a discrete specialist agent per conversation mode
 *
 * Architecture (discrete specialists):
 *
 *   Discovery agent    ─┬─ lookup_discovery_knowledge
 *                    └─ resolve_discovery
 *   Planning agent  ─┬─ lookup_planning_knowledge
 *                    └─ resolve_planning
 *   Staffing agent      ─┬─ lookup_staffing_knowledge
 *                    └─ resolve_staffing
 */

import { Agent, tool, setTracingDisabled, setDefaultOpenAIClient } from '@openai/agents';
import { AzureOpenAI } from 'openai';
import { getBearerTokenProvider } from '@azure/identity';
import { z } from 'zod';
import type { Config } from './config.ts';
import type { Logger } from './openai-client.ts';
import { createAzureCredential } from './azure-credential.ts';
import { lookupStaffingKnowledge, lookupDiscoveryKnowledge, lookupPlanningKnowledge } from './knowledge-base.ts';

// ---------------------------------------------------------------------------
// Resolution tool names (must match what prompts tell agents to call)
// ---------------------------------------------------------------------------

/** Resolution tool names — used to detect resolution from SDK output items. */
export const RESOLUTION_TOOLS = new Set(['resolve_discovery', 'resolve_planning', 'resolve_staffing']);

/** Specialist agent-tool names — emit tool.called/tool.done SSE events for these. */
export const SPECIALIST_TOOLS = new Set(['discovery_specialist', 'planning_specialist', 'staffing_specialist']);

export type SpecialistMode = 'discovery' | 'planning' | 'staffing';

export interface SpecialistAgents {
  readonly discovery: Agent;
  readonly planning: Agent;
  readonly staffing: Agent;
}

// ---------------------------------------------------------------------------
// Resolution tool definitions
// ---------------------------------------------------------------------------

/**
 * Create the three resolution FunctionTools.  The execute handlers log and
 * return a human-readable string.  Resolution detection is done by scanning
 * SDK output items (run result / stream events), not via side-effects.
 */
function createResolutionTools(log: Logger) {
  const resolveDiscovery = tool({
    name: 'resolve_discovery',
    description:
      'Call this when the Opportunity Discovery Activity concludes. Records the qualification outcome and key signals.',
    parameters: z.object({
      fit: z.enum(['qualified', 'unqualified', 'follow_up']).describe('Qualification outcome for the opportunity'),
      signals_reviewed: z.number().describe('Number of qualification signals reviewed'),
      primary_need: z.string().describe('The most important customer need or buying signal')
    }),
    execute: async (input) => {
      log.info(
        { tool: 'resolve_discovery', result: input },
        `Discovery activity resolved: ${input.fit} after ${input.signals_reviewed} signals.`
      );
      return `Discovery activity resolved: ${input.fit} after ${input.signals_reviewed} signals.`;
    }
  });

  const resolvePlanning = tool({
    name: 'resolve_planning',
    description:
      'Call this when the Account Planning Activity concludes. Records whether the account plan should advance and the next step.',
    parameters: z.object({
      approved: z.boolean().describe('Whether the plan should advance now'),
      focus_area: z.string().describe('Name of the planning'),
      next_step: z.string().describe('The next milestone, meeting, or workstream')
    }),
    execute: async (input) => {
      log.info(
        { tool: 'resolve_planning', result: input },
        `Planning activity resolved: ${input.approved ? 'advance' : 'hold'} "${input.focus_area}" with next step ${input.next_step}.`
      );
      return `Planning activity resolved: ${input.approved ? 'advance' : 'hold'} "${input.focus_area}" with next step ${input.next_step}.`;
    }
  });

  const resolveStaffing = tool({
    name: 'resolve_staffing',
    description:
      'Call this when the staffing activity concludes. Records the recommended staffing coverage and owner role.',
    parameters: z.object({
      coverage_level: z.string().describe('The recommended coverage level'),
      role: z.string().describe('The recommended owner role'),
      team_name: z.string().describe('The fictional account team name')
    }),
    execute: async (input) => {
      log.info(
        { tool: 'resolve_staffing', result: input },
        `Staffing activity resolved: ${input.coverage_level} coverage with ${input.role} on ${input.team_name}.`
      );
      return `Staffing activity resolved: ${input.coverage_level} coverage with ${input.role} on ${input.team_name}.`;
    }
  });

  return { resolveDiscovery, resolvePlanning, resolveStaffing };
}

function createKnowledgeTools(log: Logger) {
  const lookupDiscovery = tool({
    name: 'lookup_discovery_knowledge',
    description: 'Retrieve sample discovery guidance and qualification cues for the discovery specialist.',
    parameters: z.object({
      query: z.string().describe('What discovery signal, qualification detail, or customer need is needed')
    }),
    execute: async ({ query }) => {
      const matches = lookupDiscoveryKnowledge(query);
      log.info({ tool: 'lookup_discovery_knowledge', query, matches: matches.length }, 'Discovery knowledge lookup');
      return JSON.stringify({ items: matches });
    }
  });

  const lookupPlanning = tool({
    name: 'lookup_planning_knowledge',
    description: 'Retrieve sample planning guidance, priorities, and milestones for the planning specialist.',
    parameters: z.object({
      query: z.string().describe('What planning focus area, risk, or next-step detail is needed')
    }),
    execute: async ({ query }) => {
      const matches = lookupPlanningKnowledge(query);
      log.info({ tool: 'lookup_planning_knowledge', query, matches: matches.length }, 'Planning knowledge lookup');
      return JSON.stringify({ items: matches });
    }
  });

  const lookupStaffing = tool({
    name: 'lookup_staffing_knowledge',
    description: 'Retrieve sample staffing roles, coverage guidance, and qualifications for the staffing specialist.',
    parameters: z.object({
      query: z.string().describe('What staffing role, coverage, or qualification detail is needed')
    }),
    execute: async ({ query }) => {
      const matches = lookupStaffingKnowledge(query);
      log.info({ tool: 'lookup_staffing_knowledge', query, matches: matches.length }, 'Staffing knowledge lookup');
      return JSON.stringify({ items: matches });
    }
  });

  return { lookupDiscovery, lookupPlanning, lookupStaffing };
}

// ---------------------------------------------------------------------------
// Azure client setup
// ---------------------------------------------------------------------------

/**
 * Configure the OpenAI Agent SDK to use Azure OpenAI.
 *
 * The SDK uses a global default client set via `setDefaultOpenAIClient()`.
 * For Azure, we create an `AzureOpenAI` instance with a bearer token
 * provider from `@azure/identity`.
 *
 * Authentication modes:
 *   - HTTP endpoint + skipAuth (local dev / CI): uses a static dummy token
 *   - HTTPS endpoint (production): uses the runtime-appropriate Azure credential for real
 *     Azure AD tokens via the cognitiveservices scope
 */
export async function setupAzureClient(config: Config): Promise<void> {
  const isHttpEndpoint = config.azureEndpoint.startsWith('http://');
  let tokenProvider: () => Promise<string>;

  if (isHttpEndpoint && config.skipAuth) {
    tokenProvider = () => Promise.resolve('dummy');
  } else {
    const credential = createAzureCredential();
    tokenProvider = getBearerTokenProvider(credential, 'https://cognitiveservices.azure.com/.default');
  }

  const client = new AzureOpenAI({
    azureADTokenProvider: tokenProvider,
    apiVersion: config.apiVersion,
    endpoint: config.azureEndpoint
  });

  setDefaultOpenAIClient(client);
}

// ---------------------------------------------------------------------------
// Agent hierarchy creation
// ---------------------------------------------------------------------------

/**
 * Create the three specialist agents used by the sample.
 *
 * Each specialist is a user-facing conversational agent for a single mode.
 */
export function createSpecialistAgents(config: Config, log: Logger): SpecialistAgents {
  setTracingDisabled(true);

  const { resolveDiscovery, resolvePlanning, resolveStaffing } = createResolutionTools(log);
  const { lookupDiscovery, lookupPlanning, lookupStaffing } = createKnowledgeTools(log);

  const sharedInstructions = config.sharedInstructions.trim();
  const compose = (specialistInstructions: string) => `${sharedInstructions}\n\n${specialistInstructions}`.trim();

  const discoveryAgent = new Agent({
    name: 'Discovery',
    instructions: compose(config.discoveryInstructions),
    model: config.model,
    modelSettings: {
      toolChoice: 'auto'
    },
    tools: [lookupDiscovery, resolveDiscovery]
  });

  const planningAgent = new Agent({
    name: 'Planning',
    instructions: compose(config.planningInstructions),
    model: config.model,
    modelSettings: {
      toolChoice: 'auto'
    },
    tools: [lookupPlanning, resolvePlanning]
  });

  const staffingAgent = new Agent({
    name: 'Staffing',
    instructions: compose(config.staffingInstructions),
    model: config.model,
    modelSettings: {
      toolChoice: 'auto'
    },
    tools: [lookupStaffing, resolveStaffing]
  });

  log.info({ model: config.model, agentName: config.agentName }, 'Agent hierarchy initialised');

  return {
    discovery: discoveryAgent,
    planning: planningAgent,
    staffing: staffingAgent
  };
}
